### DISCUSSION TITLE
License comments in final bundle

### DISCUSSION THREAD

**Discussion by harryhjsh** [parcel-bundler/parcel#6003] [Q&A]
I'm struggling with making sure that license headers for all imported libraries are actually being included in my final bundle. 
I upgraded to v2 today partly to try and solve this problem in an easier manner, but can't get my head around what's required to make sure that license headers are actually inserted when bundled.

I'm just running `parcel build <entry> --dist-dir <dir>` with no additional options on 2.0.0-beta.2. 

With some libraries, this works fine - e.g. [SortableJS's](https://github.com/SortableJS/Sortable) license comment is preserved as it's a JSDoc comment with `@license` in its body, so terser keeps it, presumably due to [the default](https://github.com/terser/terser#format-options) `{comments: "some"}` option. (#2459 is relevant here)

However, when moving to a different project which uses React, I'm not finding myself so lucky. The React license headers are also standard JSDoc comments with `@license` in their body, but they're not getting preserved by default - they are, however, maintained in the bundle when also passing `--no-scope-hoist` as a CLI argument.

This seems similar to #3322, but given that I can't find much other mention of anything similar in the issues, I'll assume I'm missing something obvious.

 The same is true even if running with `--no-optimize` - I get parcel's `//ASSET` comments, comments from my own source files, and comments from the relevant bits of React and object-assign that are included in the bundle - but no license header.
Logically this makes some sense, at least with my (limited) understanding of how scope hoisting works in this context.
My question basically is how I can ensure that I'm passing the correct license comments into the bundle, and whether that solution is compatible with scope hoisting. It doesn't make a huge difference in this instance (the difference between a scope hoisted and non-scope-hoisted bundle in this case is 10kb or so), but I'd like to understand what's going on slightly better.

For clarity, I'm importing both libraries mentioned here with ES6 static `import` statements - e.g. `import React, { useState } from 'react';`, `import ReactDOM from 'react-dom';` and, in a different project/bundle, `import Sortable from 'sortablejs';` - I get license headers from Sortable in its bundle, but not from React or ReactDOM in its bundle.

Forgive me if this is really idiotic.

Thanks!

**Comment by FDiskas** [comment]
Only comments with 
```
/**! 
```
Are not removed. So react license comment does not have `!`
See in the bundle https://unpkg.com/browse/react@17.0.1/cjs/react-jsx-runtime.production.min.js

But sortablejs has https://unpkg.com/sortablejs@1.13.0/Sortable.js

And yes probably parcel should improve this according to the https://jsdoc.app/tags-license.html
But also be aware of multiple same license comments flooded all over the bundler.

Parcel should preserve copyright comments in the output. By default this should work like Google Closure, keeping JSDoc-style comments that contain "@license" or "@preserve". 

Would be nice to have option to pass one of the following settings: - "all" to keep all comments - `false` to omit comments in the output - and valid JS RegExp like `/foo/` or `/^!/` to keep only matching comments

There some cases that all comments are needed within bundle

**Reply by mischnic** [reply]
> Parcel should

Unless there's a bug in Parcel, all of the comment removal is handled by Terser (and there's probably a setting in `.terserrc` for comments)

**Reply by FDiskas** [reply]
So in that case according to their readme its possible
https://github.com/terser/terser#command-line-options

**Reply by harryhjsh** [reply]
The React license comment (or any other `/**` comment including `@license`) is preserved when not using scope hoisting, as the terser [format options](https://github.com/terser/terser#format-options) suggest it should be, and `.terserrc` is also respected fully when *not* using scope hoisting. (#1796 relevant wrt `.terserrc`)

I don't think I explained myself clearly enough in the OP, so I'll describe the behaviour I'm observing in brief (even when running with `--no-optimize`) - when not using scope hoisting everything functions as expected. When scope hoisting is enabled, not all comments from source files are preserved (presumably because they're not actually deemed as relevant to the final bundle by the scope hoisting process, but I could be misunderstanding this). 

If the only thing that should ever be doing comment removal is the optimizer step, I guess this is a bug if repeatable? Can anyone reliably include license headers/have a `.terserrc` fully respected with scope hoisting enabled?

**Reply by mischnic** [reply]
> scope hoisting is enabled, not all comments

> this is a bug if repeatable

Yes.

Can you open an issue with a code sample where running `--no-optimize` is missing some comments?

**Reply by harryhjsh** [reply]
Will get a minimal code sample (i.e. not including React) and do so. Thanks!
